Improve performance of discovery consumer when there are many items to index

toniprieto opened the pull request:
## Description

When there are many items to reindex using a context with the READ_ONLY mode can greatly improve indexing performance. This PR add a change to use a READ_ONLY mode during the reindexing made by discovery consumer after a commit.

## Instructions for Reviewers

List of changes in this PR:

* Change the Context mode to READ_ONLY before the indexing done by discovery consumer.
* Add a new function to do a manual flush of database session before changing the Context mode. This is necessary to ensure that we index the current object.

This could be tested by running a batch process that modifies 1000 or more items in a database transaction (filter-media, a curation task, ...)

Run the process and check the performance.

I have used this curation task made for testing (it adds and removes dc.language.iso metadada). If I run it without the change against 5000 items it takes more than 30 minutes, using the change it takes 2 minutes:

https://github.com/toniprieto/DSpace/commit/f7d790d2074b281c74fa82ff997e47e39e5d21fd

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

toniprieto committed:
Change the database mode to READ_ONLY during the indexing by discovery consumer (IndexEventConsumer)

toniprieto committed:
Add functions to do a manual flush of the db session and call flush before change to READ_ONLY mode to be sure we index the current object

mwoodiupui commented:
wonders whether #8932 might be useful in such cases.

alanorth commented:
Thanks @toniprieto. I tested this on DSpace 7.6.1 by importing a CSV with metadata changes to 1,300 items:

- Before: 235 seconds
- After: 51 seconds

This appears to drastically speed up batch indexing operations.

toniprieto commented:
@KevinVdV  Thanks for the suggestion and sorry for the late reply. 

It is true that it is weird to do the `flushSession()` call inside the discovery consumer, I’m fine with moving it although I'm not sure where to place the call to be sure that the changes made by others consumers are indexed properly. I would probably move the call to the `dispatch()` method [here](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/event/BasicDispatcher.java#L117) just before the calls to `end()` methods of the consumers. BTW, I have seen that in some cases events are dispatched without a commit.

Also, I was considering another alternative: do a flush Session inside the [setMode()](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/core/Context.java#L786-L815) method when the mode is changed to READ_ONLY. I think that this flush is only needed when READ_ONLY mode is used in case an entity is reloaded. This solution it's a way to be sure that we always use a non-dirty session in READ_ONLY mode but I don't know if it can have any side effect that I haven't taken into account.

I really appreciate your feedback. I would like to know which option you would choose.

alanorth commented:
This is a major pain point in larger repositories since DSpace 6 introduced Hibernate. I'm currently importing a CSV with changes to 5,000 items on DSpace 6.3 and it takes around _eighteen hours_ to finish.

In [2020-02 I captured some flame graphs of the Discovery indexing process on DSpace 5.10 versus 6.4-SNAPSHOT](https://alanorth.github.io/cgspace-notes/2020-02/#2020-02-09) and the difference was incredible. Line length indicates time, height indicates call stack:

DSpace 5.10:

![out dspace510-3](https://github.com/DSpace/DSpace/assets/191754/a91607c9-3568-4f3e-a06f-eb3b51d2bf79)

DSpace 6.4-SNAPSHOT:

![out dspace64-3](https://github.com/DSpace/DSpace/assets/191754/4dc584a5-0442-46fb-b232-fe257f7d516f)

Anything we can do to restore the pre-hibernate performance for the Discovery consumer would be a massive improvement to the user experience.

toniprieto commented:
Hi @alanorth 
Related with your last comment, the main/7x branches already include some improvements that have not been applied to 6.x that greatly improve the performance of indexing and batch CSV import:

https://github.com/DSpace/DSpace/pull/8046 : I applied to a 6.x instance and works ok, and improve a lot the discovery index performance
https://github.com/DSpace/DSpace/pull/8605 : I haven't tested this on 6.x, but it might be useful if you need to do more large CSV imports.

toniprieto committed:
Flush database changes after switching to READONLY mode

toniprieto committed:
Add test to check retrieving of policies after changing mode to READ_ONLY

toniprieto commented:
I just pushed changes to implement the second alternative that I mention in [this comment.](https://github.com/DSpace/DSpace/pull/8980#issuecomment-1708106525) Perform a flush in the `setMode()` method after switching to `READ_ONLY` mode.

After further analyzing [the test that initially failed and was fixed by adding the call to `flushSession()`](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/test/java/org/dspace/discovery/DiscoveryIT.java#L345-L397) I have seen that the problem in this test was a call to get the current policies of a new collection. The call, `AuthorizeService.getPoliciesActionFilter(context, dso, Constants.READ)`, which should return the READ anonymous policy, did not return any policies because the data is requested directly from the database and this row is in memory. So my understanding is that a flush is needed right after switching to `READ_ONLY` mode to ensure that the current data is always retrieved.

I added a new test to demonstrate this issue in a new `ContextModeIT` class, it should fail in the current main branch.

tdonohue commented on dspace-api/src/test/java/org/dspace/core/ContextModeIT.java:
```diff
@@ -0,0 +1,42 @@
+import org.dspace.builder.CommunityBuilder;
+import org.junit.Test;
+
+public class ContextModeIT extends AbstractIntegrationTestWithDatabase {
```
I feel we should rename this to be `ContextIT` as it's the building block for additional ITs for the Context.java class (if necessary).

tdonohue reviewed (commented):
@toniprieto : The overall concept here seems reasonable.  However, I wanted to note that I ran the new `ContextModeIT` test against `main` and **it passes there as well**.  So, I'm not sure this is a sufficient test to prove that the current behavior in the `Context` class is "wrong". 

Based on what I see in your code, I don't have objections to your approach.  But, I'm worried we may want to either test this more heavily (to ensure it provides the performance improvements *and* doesn't impact anything else), or find a better Integration Test that can prove the current behavior is incorrect.

I'll leave this backend PR in place on my local test setup for a few days and see if I notice issues.  So far it *seems* to work great.  But, I'm just hesitant in quickly merging a change to `Context` without more testing.  I'll let you know if I notice anything as I test further.  Currently I'm a 0 vote on this PR, but with more testing I think I'll likely become a +1.

One other minor note inline below.

toniprieto committed:
Change class name to ContextIT and correct a test

toniprieto commented:
Thanks for the review @tdonohue 

I have renamed the class and modified the test so that it actually fails on the main branch. I just realized that I actually made the test fail with some modifications made to `IndexEventConsumer` and that it is necessary to disable indexing during this test for it to fail on the current main branch.

tdonohue reviewed (approved):
👍 Thank you @toniprieto !  I've verified that the new `ContextIT` **fails** on `main`, but succeeds in this PR.  So, it is a good test of these new changes.    I've also done more manual testing, and I've not found any issues...everything works well and seems speedy.  The code itself looks correct to me & makes sense.

So, +1 from me.  It looks great!   Because this is modifying a core class in `Context`, I will leave this open for a few more days for others to comment on and then merge it later this week.

tdonohue commented:
Merging, as this has had a few testers who've approved.  I've found no issues in running this locally.  Thanks again @toniprieto !
